### PR TITLE
Migrate to AWS SDK v2.

### PR DESCRIPTION
**Pull request by anuraaga**
@vlsi Thanks for the work on this plugin :) I noticed it's still using AWS SDK v1 and figured it'd be good to migrate to v2. V2 is the currently developed version, and it has a netty-based client. I considered adding tcnative as well for less garbage when TLS decrypting large payloads but left that out for now, let me know if you have any thoughts on that.

### CODE REVIEW

File: src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheServiceFactory.kt [hunk 1]
```diff
@@ -69,51 +70,46 @@ class AwsS3BuildCacheServiceFactory : BuildCacheServiceFactory<AwsS3BuildCache>
...
                 }
             }
         }
     }
 
-    private fun AmazonS3ClientBuilder.addCredentials(config: AwsS3BuildCache) {
+    private fun S3AsyncClientBuilder.addCredentials(config: AwsS3BuildCache) {
         val credentials = when {
-            config.awsAccessKeyId.isNullOrBlank() && config.awsSecretKey.isNullOrBlank() -> when {
+            config.awsAccessKeyId.isNullOrBlank() || config.awsSecretKey.isNullOrBlank() -> when {
```
Comment (anuraaga) [inline src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheServiceFactory.kt:100]: Noticed I've changed the conditions here. If either key is missing, we can't provide a credential so `||` seems good here. And below, I couldn't see any reason to check whether access key is empty when creating basic creds, it seems to make more sense to check session token.

File: src/test/kotlin/com/github/burrunan/s3cache/RemoteCacheTest.kt [hunk 2]
```diff
@@ -58,7 +58,7 @@ class RemoteCacheTest: BaseGradleTest() {
         private fun gradleVersionAndSettings(): Iterable<Arguments> {
             if (!isCI) {
                 // Use the current Gradle version to make the test faster
-                return listOf(arguments("4.1"))
+                return listOf(arguments("6.6"))
```
Comment (anuraaga) [inline src/test/kotlin/com/github/burrunan/s3cache/RemoteCacheTest.kt:61]: This is now the current Gradle version as of writing :)

Comment (vlsi) [reply src/test/kotlin/com/github/burrunan/s3cache/RemoteCacheTest.kt:61]: The comment above `4.1` is not quite right.
The idea here is to use the minimal supported Gradle version for local testing, and CI adds much more versions (see below)
That enables fast local tests, and it is very easy to capture unexpected dependencies on Gradle 6 API

File: build.gradle.kts [hunk 3]
```diff
@@ -58,12 +58,13 @@ kotlinDslPluginOptions {
...
 dependencies {
-    implementation("com.amazonaws:aws-java-sdk-s3:1.11.751")
+    implementation(platform("software.amazon.awssdk:bom:2.13.76"))
+    implementation("software.amazon.awssdk:s3")
 
     testImplementation(platform("org.junit:junit-bom:5.7.0-M1"))
     testImplementation("org.junit.jupiter:junit-jupiter-api")
     testImplementation("org.junit.jupiter:junit-jupiter-params")
-    testImplementation("com.adobe.testing:s3mock-junit5:2.1.22") {
+    testImplementation("com.adobe.testing:s3mock-junit5:2.1.24") {
```
Comment (vlsi) [inline build.gradle.kts:67]: This should probably go in another PR as the update can probably remove some of the workarounds in `gradle-s3-build-cache` code.

**Comment by vlsi** [comment]
Just for the reference,

v1: 6MiB

```
1004751 aws-java-sdk-core-1.11.751.jar
 592111 aws-java-sdk-kms-1.11.751.jar
1048358 aws-java-sdk-s3-1.11.751.jar
 335042 commons-codec-1.11.jar
  61829 commons-logging-1.2.jar
 774384 httpclient-4.5.9.jar
 326874 httpcore-4.4.11.jar
 565410 ion-java-1.0.2.jar
  46968 jackson-annotations-2.6.0.jar
 258919 jackson-core-2.6.7.jar
1166637 jackson-databind-2.6.7.3.jar
  48468 jackson-dataformat-cbor-2.6.7.jar
  27585 jmespath-java-1.11.751.jar
 621931 joda-time-2.8.1.jar
6879267
```

v2: 11MiB

```
   7451 annotations-2.13.76.jar
  65384 apache-client-2.13.76.jar
  10901 arns-2.13.76.jar
 129578 auth-2.13.76.jar
  90511 aws-core-2.13.76.jar
  64422 aws-query-protocol-2.13.76.jar
  87314 aws-xml-protocol-2.13.76.jar
 335042 commons-codec-1.11.jar
  61829 commons-logging-1.2.jar
  30193 eventstream-1.0.1.jar
  60942 http-client-spi-2.13.76.jar
 774384 httpclient-4.5.9.jar
 326874 httpcore-4.4.11.jar
  68083 jackson-annotations-2.10.4.jar
 349273 jackson-core-2.10.4.jar
1404874 jackson-databind-2.10.4.jar
  19193 metrics-spi-2.13.76.jar
 278511 netty-buffer-4.1.46.Final.jar
 319530 netty-codec-4.1.46.Final.jar
 609226 netty-codec-http-4.1.46.Final.jar
 457252 netty-codec-http2-4.1.46.Final.jar
 622224 netty-common-4.1.46.Final.jar
 453883 netty-handler-4.1.46.Final.jar
 254533 netty-nio-client-2.13.76.jar
  21842 netty-reactive-streams-2.0.4.jar
  34327 netty-reactive-streams-http-2.0.4.jar
  33155 netty-resolver-4.1.46.Final.jar
 473233 netty-transport-4.1.46.Final.jar
 151888 netty-transport-native-epoll-4.1.46.Final-linux-x86_64.jar
  32965 netty-transport-native-unix-common-4.1.46.Final.jar
  26483 profiles-2.13.76.jar
  30001 protocol-core-2.13.76.jar
  11369 reactive-streams-1.0.3.jar
 426554 regions-2.13.76.jar
2513971 s3-2.13.76.jar
 523501 sdk-core-2.13.76.jar
  41117 slf4j-api-1.7.28.jar
 103805 utils-2.13.76.jar
11305618
```

File: src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt [hunk 4]
```diff
@@ -148,50 +148,72 @@ class AwsS3BuildCacheService internal constructor(
...
-            s3.getObject(bucketName, bucketPath).use { s3Object ->
-                val contentLength = s3Object.objectMetadata.contentLength
-                if (contentLength > maximumCachedObjectLength) {
+            s3Object = s3.getObject(GetObjectRequest.builder().bucket(bucketName).key(bucketPath).build(), AsyncResponseTransformer.toBytes())
+                    .join()
+        } catch (e : CompletionException) {
+            val cause = e.cause
+            when (cause) {
+                is NoSuchBucketException -> {
+                    logger.info("Bucket '{}' not found", bucketName, cause)
```
Comment (vlsi) [inline src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt:160]: `NoSuchBucketException` should be rethrown, so Gradle recognizes there's a configuration issue, and it would disable the remote cache.

File: src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt [hunk 5]
```diff
@@ -148,50 +148,72 @@ class AwsS3BuildCacheService internal constructor(
...
         val bucketPath = key.getBucketPath()
+        logger.info("Loading cache entry '{}' from S3 bucket", bucketPath)
+        val s3Object: ResponseBytes<GetObjectResponse>
         try {
-            logger.info("Loading cache entry '{}' from S3 bucket", bucketPath)
-            s3.getObject(bucketName, bucketPath).use { s3Object ->
-                val contentLength = s3Object.objectMetadata.contentLength
-                if (contentLength > maximumCachedObjectLength) {
+            s3Object = s3.getObject(GetObjectRequest.builder().bucket(bucketName).key(bucketPath).build(), AsyncResponseTransformer.toBytes())
+                    .join()
```
Comment (vlsi) [inline src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt:155]: Unfortunately, this has two issues:

1) It does receive and buffer the full response before the comparison of `contentLength > maximumCachedObjectLength`. That is it downloads full response instead of rejecting too large responses.
2) It does buffer the full blob to a byte array which might cause `OutOfMemoryError` in case the entry is big. Previous implementation streamed the response to the Gradle stream.

**Comment by vlsi** [comment]
@anuraaga, thanks for the PR!

In general, I'm ok with switching to V2 API even though it brings much more dependencies. However, I would like to keep the streaming feature, so the cache should not induce OutOfMemory in case large entries are sent/received.
Do you think you can fix it?

File: src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt (hunk 5 above)
Comment (anuraaga) [reply src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt]: Sorry for not noticing this! I have pushed a proposal commit, but it adds a ton of complexity to interface `Input/OutputStream` with the netty-based client's async API. I'm ok with reverting this commit and switching to the apache-based sync client, which presents `Input/OutputStream` natively, getting netty's efficient buffers would have been nice but not a dealbreaker. Let me know what you think.

FWIW, I have written an interface on the `putObject` side as well so that is now streamed appropriately. Maybe I can remove the reflection-based `file` optimization with it if we do go ahead and proceed with this change.

Comment (vlsi) [reply src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt]: reflection-based `file` optimization helps a lot: internally Gradle prepares the cache entry to a file on a local filesystem.
It makes it much more efficient to send that file to S3 as AWS API can retry the upload, and it can even split the file in chunks and upload it in parallel.

Comment (vlsi) [reply src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt]: I believe `OutputStream` is a wrong abstraction for cache upload:
1) It does not allow retries
2) It does not allow for sending the contents in chunks

I believe one day we could get `file()` as a public API from Gradle :)

Comment (anuraaga) [reply src/main/kotlin/com/github/burrunan/s3cache/internal/AwsS3BuildCacheService.kt]: Ah thanks - that makes sense. So I guess rather than the `OutputStream` bridge I'll revert that to the memory copy approach since it's not used much in practice anyways.

For load, any suggestion on whether to stick with this async netty API, bridging to `InputStream` or sticking with sync apache API? I'm ok with either, the bridge is not so bad but pretty complex.

**Comment by apramana** [comment]
Hello, I am interested in adding support for [S3 Transfer Acceleration](https://docs.aws.amazon.com/AmazonS3/latest/dev/transfer-acceleration-examples.html#transfer-acceleration-examples-java) - is this something that could be added in this PR? Or should this be implemented beforehand?

**Comment by juddgaddie** [comment]
This is a very old PR, is it ready to be merged ?

**Comment by vlsi** [comment]
Frankly speaking, I am inclined to sync client. Will rework to v2 sync soon

**Comment by vlsi** [comment]
@anuraaga , thanks for the PR. However, I am not confident in supporting reactive-based implementations, and it looks like the current Gradle API is not suitable for async processing.

So I reworked the PR for sync APIs in #10 which I am inclined to commit shortly.

**Comment by anuraaga** [comment]
Thanks @vlsi - yeah I think it makes sense, one reason I was hoping to use Netty is that since these are large blobs, less heap garbage would be nice. But the Gradle API being synchronous makes it a bit too complicated indeed.

**Comment by vlsi** [comment]
Thanks, merged in #10
